src/vumi2/cli.py

```diff
@@ -66,9 +66,13 @@ def worker_config_options(
     """
     for field in fields(cls):
         # Check if nested
+
+        # FIXME: The is_attrs() check now tells mypy that we have
+        # Type[AttrsInstance] instead of whatever type we actually have. Can we
+        # check for Type[BaseConfig] instead?
         if field.type and is_attrs(field.type):
```

Comment:
I've done some digging, and the types here are actually not correct. We're almost certainly getting `BaseConfig` (or a subclass thereof) in the initial `worker_config_options()` call, but `BaseConfig.amqp` has type `AmqpConfig` which is a standalone `attrs` class.

One option for fixing this is to add a new parent class to both `BaseConfig` and `AmqpConfig` whose only purpose is being at the root of the type hierarchy. While that's explicit about which things are configs and which aren't, the only place we actually care about the type is when we're building CLI options.

Another option is to change the type signature from `cls: Type[BaseConfig]` to `cls: Type[AttrsInstance]` and maybe add a wrapper that requires `BaseConfig` for the outermost call.

A third option would be to factor out the recursive walking (which we do twice more in `vumi.config`) into its own function and pass in a handler for each field in the various places we use it. (The types for this would probably end up looking like the second option.)

My preference is the third, but that belongs in its own PR. We can implement the second here as the simplest fix for the type checking.

Reply:
Yeah I'm happy with that. The main reason for having `Type[BaseConfig]` here is to document what to pass it when it's being called, so as long as we can retain that.